File: data_structures/ruby/suffix_tree.rb
class SuffixNode
    def initialize
        @value = ""
        @leaves = []
        @nodes = []
    end

    def check_nodes(suffix)
        @nodes.each do |node|
            if node.value == suffix[0]
                node.add_suffix(suffix.slice[1...suffix.length]
                return true
            end
        end
        return false
    end

    def check_leaves(suffix)
        @leaves.each do |leaf|
            if leaf[0] == suffix[0]
                node = SuffixNode.new
                node.value = leaf[0]
                node.add_suffix(suffix.slice[1...suffix.length])
                node.add_suffix(leaf.slice[1...leaf.length])
                @nodes.push(node)
                @leaves.delete_at(@leaves.index(leaf))
                return
            end
        end
        @leaves.push(suffix)
    end

    def add_suffix(suffix)
        if !suffix.empty?
            if !check_nodes(suffix)
                check_leaves(suffix)
            end
        end
        return
    end

    def get_longest_repeated_substring
        result = ""
        temp = ""
        @nodes.each do |node|
            temp = node.get_longest_repeated_substring
            if (temp.length > result.length)
                result = temp
            end
        end
        return @value + string
    end
end

class SuffixTree
    def initialize(base)
        @root = SuffixNode.new
        add_suffix(base)
    end

    def add_suffix(suffix)
        suffix.each_with_index do |char, i|
            @root.add_suffix(suffix.slice(i...suffix.length))
        end
    end

    def get_longest_repeated_substring
        @root.get_longest_repeated_substring
    end

    def contains?(substring)
        if @leaves.include?(substring)
            return true
        else
            @nodes.each do |node|
                if substring.slice(0...node.value.length) == node.value
                    if node.contains?(substring.slice(node.value.length...substring.length))
                        return true
                    end
                end
            end
        end
        return false
    end
end

File: data_structures/js/SuffixTree.js
class SuffixNode {
    constructor() {
        this.value = '';
        this.leaves = [];
        this.nodes = [];
    }

    checkNodes(suffix) {
        let node;
        for (let i = 0; i < this.nodes.length; i++) {
            node = this.nodes[i];
            if (node.value === suffix[0]) {
                node.addSuffix(suffix.slice(1));
                return true;
            }
        }
        return false;
    }

    checkLeaves(suffix) {
        let node, leaf;
        for (let i = 0; i < this.nodes.length; i++) {
            leaf = this.leaves[i];
            if (leaf[0] === suffix[0]) {
                node = new Node();
                node.value = leaf[0];
                node.addSuffix(suffix.slice(1));
                node.addSuffix(leaf.slice(1));
                this.nodes.push(node);
                this.leaves.splice(i, 1);
                return;
            }
        }
        this.leaves.push(suffix);
    }

    addSuffix(suffix) {
        if (suffix.length) {
            if (!this.checkNodes(suffix)) {
                this.checkLeaves(suffix);
            }
        }
        return;
    }

    getLongestRepeatedSubstring() {
        let str = '';
        let temp = '';
        for (let i = 0; i < this.nodes.length; i++) {
            temp = this.nodes[i].getLongestRepeatedSubstring();
            if (temp.length > str.length) {
                str = temp;
            }
        }
        return this.value + str;
    }
}

class SuffixTree {
    constructor(base) {
        this.root = new SuffixNode();
        this.addSuffix(base);
    }

    //O(length of suffix)
    addSuffix(suffix) {
        for (let i = 0; i < base.length; i++) {
            this.root.addSuffix(str.substring(i));
        }
    }

    //O(number of nodes in tree)
    getLongestRepeatedSubstring() {
        return this.root.getLongestRepeatedSubstring();
    }

    //O(number of nodes in tree)
    containsSubstring(substring) {
        if (this.leaves.incudes(substring)) {
            return true;
        } else {
            let node;
            for (let i = 0; i < this.nodes.length; i++) {
                node = this.nodes[i];
                if (substring.substring(0, node.length) === node.value) {
                    if (node.containsSubstring(substring.slice(node.length))) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
}